Update webapp template to use top-level statements & minimal hosting API

DamianEdwards opened the pull request:
Updated project templates to use top-level statements and minimal hosting APIs.

Here's what `Program.cs` ends up looking like with individual auth. (_Note that most of the `using` statements will get removed when the implicit global usings SDK PR is merged_):
``` csharp
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Builder;
using Microsoft.AspNetCore.Hosting;
using Microsoft.AspNetCore.Identity;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Hosting;
using Microsoft.Extensions.Logging;
using webapp.Data;

var builder = WebApplication.CreateBuilder(args);

// Add services to the container.
builder.Services.AddDbContext<ApplicationDbContext>(options =>
    options.UseSqlite(
        builder.Configuration.GetConnectionString("DefaultConnection")));
builder.Services.AddDatabaseDeveloperPageExceptionFilter();
builder.Services.AddDefaultIdentity<IdentityUser>(options => options.SignIn.RequireConfirmedAccount = true)
    .AddEntityFrameworkStores<ApplicationDbContext>();
builder.Services.AddRazorPages();

var app = builder.Build();

// Configure the HTTP request pipeline.
if (app.Environment.IsDevelopment())
{
    app.UseDeveloperExceptionPage();
    app.UseMigrationsEndPoint();
}
else
{
    app.UseExceptionHandler("/Error");
    // The default HSTS value is 30 days. You may want to change this for production scenarios, see https://aka.ms/aspnetcore-hsts.
    app.UseHsts();
}

app.UseHttpsRedirection();
app.UseStaticFiles();

app.UseAuthentication();
app.UseAuthorization();

app.MapRazorPages();

app.Run();
```

Contributes to #33947 #33948 #33944 #32471 #33813

DamianEdwards committed:
Update webapp template to use top-level statements & minimal hosting API

Contributes to #33947 #33944

bradygaster reviewed (approved):
👍

pranavkm commented:
@SteveSandersonMS / @javiercn in case you'd like to chime in on the formatting / shape of the code.

pranavkm commented:
```C#
builder.Services.AddDbContext<ApplicationDbContext>(options =>
    options.UseSqlite(
        builder.Configuration.GetConnectionString("DefaultConnection")));
```

Is it more legible if the connection string is stored in a variable?

```C#
var connectionString = builder.Configuration.GetConnectionString("DefaultConnection");
builder.Services.AddDbContext<ApplicationDbContext>(options => options.UseSqlite(connectionString));
```

DamianEdwards commented:
@pranavkm I'm happy to make that change. @davidfowl has some pretty strong opinions on this and did not like my original proposal that declared variables for `services` and `configuration` to avoid the `builder.Services` and `builder.Configuration` references everywhere.

DamianEdwards commented:
Note that when the global implicit namespaces change lands the namespaces in this `Program.cs` will be reduced to:
``` csharp
using Microsoft.AspNetCore.Identity; // Needed for the IdentityUser type
using Microsoft.EntityFrameworkCore; // Needed for the UseSqlite extension method
using webapp.Data; // Needed for the ApplicationDbContext type
```

pranavkm commented:
>  @davidfowl has some pretty strong opinions on this

I guess let's keep with this for now and we can always update it once @davidfowl's had a chance to weigh in on the change.

DamianEdwards committed:
Remove unused usings from WebApp template

DamianEdwards committed:
Updated StarterWeb template:
- Using minimal hosting APIs
- Using top-level statements

DamianEdwards committed:
Updated gRPC template:
- using minimal hosting APIs
- using top-level statements

DamianEdwards commented:
@JamesNK in case you'd like to chime in on the gRPC project template changes.

DamianEdwards committed:
Updated WebApi template:

- uses minimal hosting APIs
- uses top-level statements

DamianEdwards committed:
Updated BlazorServer project template:

- uses minimal hosting APIs
- uses top-level statements

JamesNK reviewed (approved):
gRPC template 👍

DamianEdwards committed:
Update BlazorWasm template

- Uses minimal hosting APIs
- Uses top-level statements
- Fixed template testing scripts to support BlazorWasm template properly
- Fixed nullable issues

pranavkm reviewed (approved):
As long as it works!

DamianEdwards committed:
Alias Microsoft.Graph namespace to fix type conflict

davidfowl commented:
There's a super subtle difference to where UseRouting is called implicitly with the WebApplicationBuilder that we should make sure is ok. Routing will run before static files in a couple of these templates. Maybe we want to change that back so it's more explicit in those scenarios?

DamianEdwards commented:
@davidfowl that's a good point but I must say I love the fact we can omit it. Is the difference big enough to warrant being explicit you think? Is the concern about performance, behavior, correctness, ease of diagnosing failure modes, etc.?

davidfowl commented:
I guess to 2 differences are:
- Routes can win over static files. This might really only matter for implicit defaults (index.htm etc)
- Performance of static files as routing runs before it.

These can be worked around obviously by adding an explicit call but maybe we should measure.

javiercn commented:
> * Performance of static files as routing runs before it.

Doesn't routing run before static files today? Matching at least will.

javiercn commented:
> * Routes can win over static files. This might really only matter for implicit defaults (index.htm etc)

This is also the case today, isn't it? The `MapFallback` overloads have explicit constraints to avoid matching things that look like files, if you register a slug route explicitly, it will win over the files.

I believe the StaticFiles middleware avoids doing work if it detects there's an endpoint set on the request, so routes should already win?

davidfowl commented:
> Doesn't routing run before static files today? Matching at least will.

No it doesn't. For example here is the Razor pages template:

```C#
public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
{
    if (env.IsDevelopment())
    {
        app.UseDeveloperExceptionPage();
    }
    else
    {
        app.UseExceptionHandler("/Error");
        // The default HSTS value is 30 days. You may want to change this for production scenarios, see https://aka.ms/aspnetcore-hsts.
        app.UseHsts();
    }

    app.UseHttpsRedirection();
    app.UseStaticFiles();

    app.UseRouting();

    app.UseAuthorization();

    app.UseEndpoints(endpoints =>
    {
        endpoints.MapRazorPages();
    });
}
```

Things that run before routing:
- DeveloperExceptionPage
- ExceptionHandler
- Hsts
- HttpsRedirection
- StaticFiles

> I believe the StaticFiles middleware avoids doing work if it detects there's an endpoint set on the request, so routes should already win?


That's not the problem the "problem" is that it runs route matching first, which is a change from today. It may be irrelevant but we should be aware of the ordering differences here.

davidfowl commented:
> This is also the case today, isn't it? The MapFallback overloads have explicit constraints to avoid matching things that look like files, if you register a slug route explicitly, it will win over the files.

I'm not sure what you mean. If you explicitly call MapFallback? Sure but I'm not talking about that scenario.

davidfowl commented on src/ProjectTemplates/Web.ProjectTemplates/content/RazorPagesWeb-CSharp/Program.cs:
```diff
@@ -2,25 +2,137 @@
+    options.FallbackPolicy = options.DefaultPolicy;
+});
+builder.Services.AddRazorPages()
+    .AddMvcOptions(options => {})
```
Is this what it does today?

DamianEdwards replied:
Yep https://github.com/dotnet/aspnetcore/blob/main/src/ProjectTemplates/Web.ProjectTemplates/content/RazorPagesWeb-CSharp/Startup.cs#L106

davidfowl replied:
Unrelated to this PR but wtf. @pranavkm @javiercn can we fix this?

pranavkm replied:
Lol, why is this here? We should be able to remove it.

```suggestion
```

davidfowl commented on src/ProjectTemplates/Web.ProjectTemplates/content/RazorPagesWeb-CSharp/Program.cs:
```diff
@@ -2,25 +2,137 @@
+#endif
+#elif (IndividualB2CAuth)
+#if (GenerateApi)
+var initialScopes = builder.Configuration.GetValue<string>("DownstreamApi:Scopes")?.Split(' ');
```
```suggestion
var initialScopes = builder.Configuration["DownstreamApi:Scopes"]?.Split(' ');
```

DamianEdwards replied:
I can make these changes but this is a straight port from what's there today.

davidfowl commented on src/ProjectTemplates/Web.ProjectTemplates/content/RazorPagesWeb-CSharp/Program.cs:
```diff
@@ -2,25 +2,137 @@
+#endif
+builder.Services.AddAuthentication(OpenIdConnectDefaults.AuthenticationScheme)
+#if (GenerateApiOrGraph)
+    .AddMicrosoftIdentityWebApp(builder.Configuration.GetSection("AzureAd"))
```
Why the nesting of dotted things at the same level?

DamianEdwards replied:
Are you referring to all the Identity methods? If so it's matching what it is today: https://github.com/dotnet/aspnetcore/blob/main/src/ProjectTemplates/Web.ProjectTemplates/content/RazorPagesWeb-CSharp/Startup.cs#L69

davidfowl replied:
ugh

davidfowl commented on src/ProjectTemplates/Web.ProjectTemplates/content/StarterWeb-CSharp/Program.cs:
```diff
@@ -2,25 +2,139 @@
+    .AddEntityFrameworkStores<ApplicationDbContext>();
+#elif (OrganizationalAuth)
+#if (GenerateApiOrGraph)
+var initialScopes = builder.Configuration.GetValue<string>("DownstreamApi:Scopes")?.Split(' ');
```
```suggestion
var initialScopes = builder.Configuration["DownstreamApi:Scopes"]?.Split(' ');
```

davidfowl commented on src/ProjectTemplates/Web.ProjectTemplates/content/StarterWeb-CSharp/Program.cs:
```diff
@@ -2,25 +2,139 @@
+
+builder.Services.AddControllersWithViews(options =>
+{
+    var policy = new AuthorizationPolicyBuilder()
```
@pranavkm @HaoK I thought we did this with middleware now. Why does this template still use the authorize filter?

DamianEdwards replied:
Yeah that looks strange. Seems like a relic?

davidfowl commented:
Not gonna lie these templates are impossible to review because of those ifs. You really want to look at the resulting content for each of them. Those could be baseline files checked into the repo or something.

DamianEdwards commented:
@davidfowl certain baselines are generated in the `scripts` folder but purposely excluded as they're not part of the build today (i.e. they wouldn't get updated unless one manually runs the various `Run-***-Locally.ps1` scripts). I can put the relevant baselines in a gist if it helps for review for now but any larger change to how template baselines are tracked is outside the scope of this work.

DamianEdwards committed:
Address feedback

DamianEdwards committed:
Remove LangVersion=preview

DamianEdwards committed:
Feedback

DamianEdwards committed:
Address feedback

- put connection string in its own variable
- add explicit call to UseRouting() to avoid issue with middleware that re-runs the pipeline with a different path (e.g. UseExceptionHandler)

DamianEdwards commented:
I've added calls to `UseRouting()` in all templates that have a call to `UseExceptionHandler` as not doing that causes issues today (see #34146).

@pranavkm @davidfowl I did the `var connectionString ...` change too.

Planning to merge once checks finish on this latest commit.
